starknet_py/net/contract/contract_function.py

```diff
@@ -0,0 +1,481 @@
+    _account: Optional[BaseAccount]
+
+    def __post_init__(self):
+        if self._account is None:
+            raise ValueError(
+                "Contract instance was created without providing an Account. "
+                "It is not possible to prepare and send an invoke transaction."
+            )
+
+    @property
+    def get_account(self):
+        if self._account is not None:
+            return self._account
+
+        raise ValueError(
+            "The account is not defined. It is not possible to send an invoke transaction."
+        )
+
+    @abstractmethod
+    async def estimate_fee(
+        self,
+        block_hash: Optional[Union[Hash, Tag]] = None,
+        block_number: Optional[Union[int, Tag]] = None,
+        *,
+        nonce: Optional[int] = None,
+    ) -> EstimatedFee:
+        """
+        Estimate fee for prepared function call.
+
+        :param block_hash: Estimate fee at specific block hash.
+        :param block_number: Estimate fee at given block number
+            (or "latest" / "pending" for the latest / pending block), default is "pending".
+        :param nonce: Nonce of the transaction.
+        :return: Estimated amount of the transaction cost, either in Wei or Fri associated with executing the
+            specified transaction.
+        """
+
+    async def _invoke(self, transaction: Invoke) -> InvokeResult:
+        response = await self._client.send_transaction(transaction)
+
+        invoke_result = InvokeResult(
+            hash=response.transaction_hash,  # noinspection PyTypeChecker
+            _client=self._client,
+            contract=self._contract_data,
+            invoke_transaction=transaction,
+        )
+
+        return invoke_result
+
+
+@add_sync_methods
+@dataclass
+class PreparedFunctionInvokeV1(PreparedFunctionInvoke):
+    max_fee: Optional[int]
+
+    async def invoke(
+        self,
+        max_fee: Optional[int] = None,
+        auto_estimate: bool = False,
+        *,
+        nonce: Optional[int] = None,
+    ) -> InvokeResult:
+        """
+        Send an Invoke transaction version 1 for a prepared data.
+
+        :param max_fee: Max amount of Wei to be paid when executing transaction.
+        :param auto_estimate: Use automatic fee estimation, not recommend as it may lead to high costs.
+        :param nonce: Nonce of the transaction.
+        :return: InvokeResult.
+        """
+
+        transaction = await self.get_account.sign_invoke_v1_transaction(
+            calls=self,
+            nonce=nonce,
+            max_fee=max_fee or self.max_fee,
+            auto_estimate=auto_estimate,
+        )
+
+        return await self._invoke(transaction)
+
+    async def estimate_fee(
+        self,
+        block_hash: Optional[Union[Hash, Tag]] = None,
+        block_number: Optional[Union[int, Tag]] = None,
+        *,
+        nonce: Optional[int] = None,
+    ) -> EstimatedFee:
+        tx = await self.get_account.sign_invoke_v1_transaction(
```

Comment:
I looked into `sign_invoke_v1_transaction`. Why aren't we using query versions for fee estimation?

I understand that it was this way done before, but since RPC 0.6.0 reintroduced query versions, maybe it would make sense to adopt a more correct approach, either now or in a separate PR in the nearby future.

Reply:
Good catch! Added `sign_for_fee_estimate()` call to `estimate_fee` methods in `PreparedFunctionInvokeV1` and `PreparedFunctionInvokeV3`.